File: neet_code_150_cpp/graphs/surrounded_regions.cpp
#include <iostream>
#include <vector>

// 模範解答
class SolutionAns {
    public:
        void solve(std::vector<std::vector<char>>& board) {
            auto [m, n] = std::pair((int)board.size(), (int)board[0].size());

            // 右端、左端
            for(int i = 0; i < m; ++i) {
                dfs(board, {i, 0});
                dfs(board, {i, n - 1});
            }

            // 上端、下端
            for(int j = 0; j < n; ++j) {
                dfs(board, {0, j});
                dfs(board, {m - 1, j});
            }

            for(int i = 0; i < m; ++i) {
                for(int j = 0; j < n; ++j) {

                    if(board[i][j] == 'O') {
                        board[i][j] = 'X';
                    }

                    if(board[i][j] == 'E') {
                        board[i][j] = 'O';
                    }
                }
            }
        }

    private:
        void dfs(std::vector<std::vector<char>>& board, std::pair<int, int> v) {
            auto [m, n] = std::pair((int)board.size(), (int)board[0].size());
            // 境界チェック
            if(v.first < 0 
                    || v.first >= m 
                    || v.second < 0 
                    || v.second >= n 
                    || board[v.first][v.second] != 'O' 
                    ) 
            {
                return;
            }

            board[v.first][v.second] = 'E';

            // 十字方向の再帰的な探索
            std::vector<std::pair<int, int>> directions = {
                {1, 0},
                {0, 1},
                {-1, 0},
                {0, -1},
            };
            for(auto& s: directions) {
                dfs(board, {v.first + s.first, v.second + s.second});
            }
        }

};

void show_result(const std::vector<std::vector<char>>& result) {
    for(auto& v: result) {
        for(auto& w: v) {
            std::cout << w << " ";
        }
        std::cout << '\n';
    }
}

int main(void) {
    std::vector<std::vector<char>> case_1 = {
        {'X', 'X', 'X', 'X'},
        {'X', 'O', 'O', 'X'},
        {'X', 'X', 'O', 'X'},
        {'X', 'O', 'X', 'X'},
    };
    // =>
    /*
       [ ['X', 'X', 'X', 'X']
         ['X', 'X', 'X', 'X']
         ['X', 'X', 'X', 'X']
         ['X', 'O', 'X', 'X']]
    */
    std::vector<std::vector<char>> case_2 = {{'X'}};
    // => [['X']]

    SolutionAns s_ans;
    auto res_1 = case_1;
    auto res_2 = case_2;

    s_ans.solve(res_1);
    s_ans.solve(res_2);

    show_result(res_1);
    show_result(res_2);
}

File: neet_code_150/src/graphs/surrounded_regions.rs
// 解けなかった
struct Solution {}
impl Solution {
    pub fn solve(board: &mut Vec<Vec<char>>) {
        let (m, n) = (board.len(), board[0].len());

        for x in 0..m {
            for y in 0..n {
                if board[x][y] == 'O' {
                    Self::dfs(board, (x as i32, y as i32));
                }
            }
        }
    }

    fn dfs(board: &mut Vec<Vec<char>>, v: (i32, i32)) {
        let (m, n) = (board.len(), board[0].len());
        // 境界チェック
        if v.0 < 0
            || v.1 < 0
            || v.0 >= m as i32
            || v.1 >= n as i32
            || board[v.0 as usize][v.1 as usize] == 'X'
        {
            return;
        }

        // 訪問済み
        board[v.0 as usize][v.1 as usize] = 'X';
        // 十字方向の再帰的な探索
        let directions: [(i32, i32); 4] = [(0, 1), (1, 0), (0, -1), (-1, 0)];
        for s in directions {
            let v_new = (v.0 + s.0, v.1 + s.1);
            Self::dfs(board, v_new);
        }
    }

    fn dfs_2(board: &mut Vec<Vec<char>>, v: (i32, i32)) -> bool {
        let (m, n) = (board.len(), board[0].len());

        // 十字方向の再帰的な探索
        let directions: [(i32, i32); 4] = [(0, 1), (1, 0), (0, -1), (-1, 0)];
        for s in directions {
            if board[v.0 as usize][v.1 as usize] == 'O' {
                let v_new = (v.0 + s.0, v.1 + s.1);
            }
        }

        true
    }
}

// AC
// C++の模範解答より
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    pub fn solve(board: &mut Vec<Vec<char>>) {
        let (m, n) = (board.len(), board[0].len());

        // 右端、左端
        for i in 0..m {
            Self::dfs(board, (i as i32, 0));
            Self::dfs(board, (i as i32, n as i32 - 1));
        }

        // 上端、下端
        for j in 0..n {
            Self::dfs(board, (0, j as i32));
            Self::dfs(board, (m as i32 - 1, j as i32));
        }

        for i in 0..m {
            for j in 0..n {
                if board[i][j] == 'O' {
                    board[i][j] = 'X';
                }
                if board[i][j] == 'E' {
                    board[i][j] = 'O';
                }
            }
        }
    }

    fn dfs(board: &mut Vec<Vec<char>>, v: (i32, i32)) {
        let (m, n) = (board.len(), board[0].len());

        // 境界チェック
        if v.0 < 0
            || v.0 >= m as i32
            || v.1 < 0
            || v.1 >= n as i32
            || board[v.0 as usize][v.1 as usize] != 'O'
        {
            return;
        }

        // 'O'を'E'に書き換える
        board[v.0 as usize][v.1 as usize] = 'E';

        // 十字方向の再帰的な探索
        let directions: [(i32, i32); 4] = [(0, 1), (1, 0), (0, -1), (-1, 0)];
        for s in directions {
            let v_new = (v.0 + s.0, v.1 + s.1);
            Self::dfs(board, v_new);
        }
    }
}

fn main() {
    let case_1 = vec![
        vec!['X', 'X', 'X', 'X'],
        vec!['X', 'O', 'O', 'X'],
        vec!['X', 'X', 'O', 'X'],
        vec!['X', 'O', 'X', 'X'],
    ];
    // =>
    /*
       [ ['X', 'X', 'X', 'X']
         ['X', 'X', 'X', 'X']
         ['X', 'X', 'X', 'X']
         ['X', 'O', 'X', 'X']]
    */
    let case_2 = vec![vec!['X']];
    // => [['X']]

    let mut res_1 = case_1.clone();
    let mut res_2 = case_2.clone();
    SolutionAnsCpp::solve(&mut res_1);
    SolutionAnsCpp::solve(&mut res_2);

    println!("res_1: {:?}", res_1);
    println!("res_2: {:?}", res_2);
}
